File: java/src/main/java/com/dokmatiq/docgen/client/ZugferdClient.java
package com.dokmatiq.docgen.client;

import com.dokmatiq.docgen.internal.FileUtils;
import com.dokmatiq.docgen.internal.Transport;
import com.dokmatiq.docgen.model.InvoiceData;

import java.nio.file.Path;
import java.util.Map;

/** Client for ZUGFeRD/Factur-X operations. */
public class ZugferdClient {
    private final Transport transport;

    public ZugferdClient(Transport transport) {
        this.transport = transport;
    }

    /** Embed ZUGFeRD XML into a PDF. */
    public byte[] embed(Path file, InvoiceData invoiceData) {
        return transport.requestBytes("POST", "/api/zugferd/embed",
                Map.of("pdfBase64", FileUtils.toBase64(file), "invoiceData", invoiceData));
    }

    /** Embed ZUGFeRD XML from bytes. */
    public byte[] embed(byte[] data, InvoiceData invoiceData) {
        return transport.requestBytes("POST", "/api/zugferd/embed",
                Map.of("pdfBase64", FileUtils.toBase64(data), "invoiceData", invoiceData));
    }

    /** Extract ZUGFeRD XML from a PDF. */
    public InvoiceData extract(Path file) {
        return transport.requestJson("POST", "/api/zugferd/extract",
                Map.of("pdfBase64", FileUtils.toBase64(file)), InvoiceData.class);
    }

    /** Validate a ZUGFeRD PDF. */
    public Map<?, ?> validate(Path file) {
        return transport.requestJson("POST", "/api/zugferd/validate",
                Map.of("pdfBase64", FileUtils.toBase64(file)), Map.class);
    }
}

File: dotnet/src/DocGen/Clients/ZugferdClient.cs
using Dokmatiq.DocGen.Internal;
using Dokmatiq.DocGen.Models;

namespace Dokmatiq.DocGen.Clients;

/// <summary>
/// Client for ZUGFeRD/Factur-X operations.
/// </summary>
public sealed class ZugferdClient
{
    private readonly Transport _transport;

    internal ZugferdClient(Transport transport) => _transport = transport;

    /// <summary>Embed ZUGFeRD XML into a PDF file.</summary>
    public byte[] Embed(string filePath, InvoiceData invoiceData)
        => _transport.RequestBytes(HttpMethod.Post, "/api/zugferd/embed/base64",
            new { pdfBase64 = FileUtils.ToBase64(filePath), invoice = invoiceData });

    /// <summary>Embed ZUGFeRD XML from bytes.</summary>
    public byte[] Embed(byte[] data, InvoiceData invoiceData)
        => _transport.RequestBytes(HttpMethod.Post, "/api/zugferd/embed/base64",
            new { pdfBase64 = FileUtils.ToBase64(data), invoice = invoiceData });

    /// <summary>Extract ZUGFeRD XML from a PDF.</summary>
    public InvoiceData Extract(string filePath)
        => _transport.RequestJson<InvoiceData>(HttpMethod.Post, "/api/zugferd/extract/base64",
            new { pdfBase64 = FileUtils.ToBase64(filePath) });

    /// <summary>Validate a ZUGFeRD PDF.</summary>
    public Dictionary<string, object> Validate(string filePath)
        => _transport.RequestJson<Dictionary<string, object>>(HttpMethod.Post, "/api/zugferd/validate/base64",
            new { pdfBase64 = FileUtils.ToBase64(filePath) });
}
